resolve .zil domains guide (#35)
unstoppabledomains/docs: src/guides/resolve-zil-domain.md

---
description: >-
  This tutorial will teach you how to make a JSON RPC call to our Zilliqa smart
  contract to get various records as well as owner address of a domain owner.
---

# Resolving .zil domain

## Steps to resolve unstoppable domain

In order to fetch any data from the Zilliqa blockchain you will need to do the following:

1. Get a namehash of a domain
2. Get resolver and owner addresses
3. Get the contract sub-state related to your domain

### Namehashing

> Namehashing is an algorithm that converts a domain name in a classical format \(like www.example.crypto\) to ERC-721 token id. This process is described  in more details over  [here](https://docs.unstoppabledomains.com/domain-registry-essentials/namehashing)

Reviewer: Let's use a more meaningful link text.
https://developers.google.com/style/cross-references

Author: Since it is a zilliqa part I have removed the link to our explanation of namehashing in order to not confuse people with ERC-721

> Namehashing is an algorithm that converts a domain name in a classical format \(like www.example.crypto\) to ERC-721 token id. This process is described  in more details over  [here](https://docs.unstoppabledomains.com/domain-registry-essentials/namehashing)

Reviewer: I think it's also important to note the difference between namehashing in ETH and ZIL. (In ZIL we use `sha256` form `SHA-2`, instead of `keccak256`).

```typescript

> import hash from 'hash.js';

Reviewer: Is it this library? https://www.npmjs.com/package/hash.js/v/1.0.1

I'd probably consider using something else, which is more common in JS. Afaik, NodeJS has its own `crypto`. Or we could either use this: https://www.npmjs.com/package/js-sha256

function namehash(domain: string): string {
    const parent =
      '0000000000000000000000000000000000000000000000000000000000000000';
    return '0x' + [parent]
      .concat(
        domain
          .split('.')
          .reverse()

>           .filter(label => label),

Reviewer: I'd probably skip this line since it doesn't look like the algorithm described in https://eips.ethereum.org/EIPS/eip-137 and can be misleading about label validation rules.

      )
      .reduce((parent, label) =>
        childhash(parent, label),
      );
  }

function childhash( parent: string, label: string): string {
    parent = parent.replace(/^0x/, '');
    return sha256(parent + sha256(label), 'hex',);
  }

function sha256(message: string, inputEnc?: 'hex') {
    return hash.sha256()
      .update(message, inputEnc)
      .digest('hex');
  }
```

> In more simplistic human words we are hashing all of the labels of the domain with the accumulated hash of its parents in order to get a unique ERC-721 token id. 

Reviewer: I don't think I like the idea of mentioning ERC-721 in an article about Zilliqa.

Reviewer 2: Agree. Namehash is not ERC-721 token id by itself. It's valid for .crypto implementation only - it's not a part of standard.

### Getting resolver and owner addresses

In order to get the desirable addresses, we would need a Zilliqa registry contract address 

| Zilliqa registry contract address | Zilliqa API url |
| :--- | :--- |
| 0x9611c53BE6D1b32058b2747bdeCECed7e1216793 | https://api.zilliqa.com/ |

Next, we are going to make a post API call to Zilliqa API with the parameters of our request

{% api-method method="post" host="https://api.zilliqa.com" path="/" %}
{% api-method-summary %}
​Fetch owner and resolver addresses
{% endapi-method-summary %}

{% api-method-description %}
For params, you will need to pass an array containing three values  
  

> \* contractAddress  -&gt;  **"0x9611c53BE6d1b32058b2747bdeCECed7e1216793"**  

Reviewer 2: Consider using markdown tables for such content

---

> description: Simplest step-by-step guide on how to resolve .zil domain

Reviewer: ```suggestion
description: A step-by-step guide on how to resolve .zil domains
```

---

> # .zil Integration

Reviewer: ```suggestion
# Resolve a .zil domain
```

I think I like the previous variant more, to avoid the dot in the beginning

In order to resolve a .zil domain, we will need to satisfy three steps below

* Get a namehash of a domain
* Get resolver contract address from a domain
* Query resolver contract to fetch the records

Let's visualize the resolution process using some of the simplest tools web developer has: knowledge of HTML and js.

### Initialize the project folder

As has been said above all we need is to create a folder and two files index.html and index.js respectively

```text
mkdir unstoppable-zil-resolution
cd unstoppable-zil-resolution
touch index.html index.js
```

> Let's use some blank HTML page. We have connected our empty index.js as well as a CDN library of [**js-sha256**](https://www.npmjs.com/package/js-sha256?utm_source=cdnjs&utm_medium=cdnjs_link&utm_campaign=cdnjs_library) for future namehashing step.

Reviewer: ```suggestion
Let's create a blank HTML page. We have connected our empty index.js as well as a CDN library of [**js-sha256**](https://www.npmjs.com/package/js-sha256) for future namehashing step.
```

> ```markup

Reviewer: ```suggestion
```html
```

<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>Basic .zil integration</title>
    </head>
    <body>

      <script 
        src="https://cdnjs.cloudflare.com/ajax/libs/js-sha256/0.9.0/sha256.min.js" 
        integrity="sha512-szJ5FSo9hEmXXe7b5AUVtn/WnL8a5VofnFeYC2i2z03uS2LhAch7ewNLbl5flsEmTTimMN0enBZg/3sQ+YOSzQ==" 

>         crossorigin="anonymous"></script>

Reviewer: We should either use 4 spaces here or 2 spaces indentation in other parts to be consistent.

      <script src="index.js"></script>
    </body>
</html>
```

Don't forget to add an input field and a button that will trigger the resolution process

```markup
      <input id="input" />
      <button onclick="resolve()">Resolve</button>
```

#### As for our index.js file

We are going to put some basic code to capture the text from the input field and print it in our console

```javascript

> // <!-------> Resolving domain <!------->

Reviewer: ```suggestion
```
The function name is pretty much self-descriptive :)

async function resolve() {
  const userInput = document.getElementById("input").value;
  console.log({ domain: userInput });
}
```

### Taking a namehash

Namehashing is an algorithm that converts a domain name in a classical format \(like example.crypto\) to a token id that Zilliqa contract can understand. 

{% hint style="warning" %}

> It is important to know the difference between Zilliqa namehashing and [ERC-721](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm) which is a part of [EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md).

Reviewer: It sounds like we compare Zilliqa namehashing to ERC-721 itself. Btw, namehashing isn't a part of the ERC-721 standard. I'd rather mention the same standard from the link.

> In ZIL we use **sha256 from SHA-2**, instead of **keccak256** which is used across Ethereum chain

Reviewer: ```suggestion
In ZIL we use **sha256 from SHA-2**, instead of **keccak256**, which is used across Ethereum.
```

{% endhint %}

To do so we need to split the domain by "." character to get each label and then reduce the label's array with a sha256 hashing of an accumulator and next label starting from the end. 

> For the purposes of keeping this tutorial short, instead of going into the details of this process, we are going to use the namehash function with some adaptation to the hashing library

Reviewer: ```suggestion
For the purposes of keeping this tutorial short, instead of going into the details of this process, we are going to use the namehash function with some adaptation to the hashing library.
```

Btw, what do you mean by the adaptation?

Author: shaWrapper function was not initially planned and I had to include it there due to cdn library limitation. the sha library on cdn doesn't really convert hexString to an array of hex values which is required for namehashing

{% hint style="info" %}
Don't forget to add the js-sha256 library to the project in order to use the sha256 function. 

> {% endhint %}

Reviewer: It's something that we instructed a reader to do when describing the creation of the HTML page. Not sure it's worth to stress it again.

```javascript

> // <!-------> Namehashing functions <!------->

Reviewer: ```suggestion
```

function namehash(domain) {
    const parent =
      '0000000000000000000000000000000000000000000000000000000000000000';
    return '0x' + [parent]
      .concat(
        domain
          .split('.')
          .reverse()
      )
      .reduce((parent, label) =>
        childhash(parent, label),
      );
  }

function childhash( parent, label) {
    parent = parent.replace(/^0x/, '');
    return shaWrapper(parent + shaWrapper(label), "hex");
  }

function shaWrapper(msg, inputEnc) {
  if (inputEnc === "hex") {
    var res = [];
    msg = msg.replace(/[^a-z0-9]+/ig, '');
      if (msg.length % 2 !== 0)
        msg = '0' + msg;
    for (var i = 0; i < msg.length; i += 2)
        res.push(parseInt(msg[i] + msg[i + 1], 16));
    return sha256(res);
  }
  return sha256(msg);

> }

Reviewer: Can we rewrite the whole `namehash` function to be a simple recursive function that returns a `Uint8Array`? (`sha256` can do it.) Then we could introduce a simple function that turns an array into a hex string (like [this](https://stackoverflow.com/a/40031979/4284929)). It would be great if we made it similar to the namehashing algorithm that we reference in the article.

The fact that we convert strings back and forth makes this code really overcomplicated.

Author: namehash and childhash functions are taken from our resolution library. shaWrapper is needed because of the sha library we use here. We are not using any bundler and limited to this CDN which doesn't allow us to define the input Encoding. On our resolution side, the shaWrapper is being internal inside the sha function and we don't see it.

I can try to put some comments around it to make it more pleasable, but I think if our users want to know more about namehashing they should go for https://docs.unstoppabledomains.com/domain-registry-essentials/namehashing article. I expect most users reading this guide to either copy the namehashing function into their project or ignore it all at once and write their own algorithm.

---
description: Simplest step-by-step guide on how to resolve .zil domain
---

# Resolve a .zil domain

## Where can I get .zil domain?

> The most simpliest way is to go to [our homepage](https://www.unstoppabledomains.com) and buy one for yourself

Reviewer 2: ```suggestion
The most simpliest way is to go to [Unstoppable Domains website](https://www.unstoppabledomains.com) and buy one for yourself
```

* [Get resolver contract address from a domain](resolve-zil-domain.md#getting-resolver-address)
* [Query resolver contract to fetch the records](resolve-zil-domain.md#fetching-the-domain-records)
* [Displaying results](resolve-zil-domain.md#displaying-result)
* [Error handling](resolve-zil-domain.md#error-handling)

## Initialize the project folder

We are going to build the project using [parcel.js](https://parceljs.org/) as it allows us to include cryptographic library [hash.js](https://www.npmjs.com/package/hash.js/v/1.1.7) for encoding the domain in the future and typescript supports from the start. For now, let's open a terminal and initialize the project.

```text
npm install -g add parcel-bundler
mkdir unstoppable-zil-resolution
cd unstoppable-zil-resolution
npm init
npm install hash.js
touch index.ts index.html

.
├── README.md
├── index.html
├── index.ts
├── node_modules
├── package.json

```

Now open **package.json** and configure run scripts for us

```javascript
  "scripts": {
    "dev": "parcel index.html",
    "build": "parcel build index.html"
  }
```

As our next step, let's create a blank HTML page. Create **index.html** and use the following code.

> 

Reviewer 2: I would put filename as a header just above the code. 

### index.html

```markup
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>Basic .zil integration</title>
    </head>
    <body>
      <div id="main" style="
        display: flex;
        flex-direction: column;
        height: 100vh;
      ">
        <input id="input" />
        <button id="button">Resolve</button>
        
        <div id="records" style="display: flex; flex-direction: column;">

        </div>
      </div>
      <script src="./index.ts"></script>
  </body>
</html>
```

{% hint style="info" %}
Take notice of this line `code <script src="./index.ts"></script>.` Parcel package will automatically convert this line to the js bundle it generates from our code.
{% endhint %}

> ### **Create** ~~**index.js**~~  **index.ts**

Reviewer 2: @JohnnyJumper 

This is how it looks right now. 
<img width="261" alt="image" src="https://user-images.githubusercontent.com/3999180/106001597-32abb300-60c1-11eb-9070-5a1589ae500d.png">

Can you edit markdown in the editor instead of gitbook? I believe vscode has great options for editing, linting and preview markdown: https://code.visualstudio.com/docs/languages/markdown

Create file index.ts. We are going to define and attach a function to our HTML button under `code id="button"` 

```typescript

> // index.ts

Reviewer 2: I would put filename as a header just above the code. 

### index.ts

async function resolve() {
  const userInput = (<HTMLInputElement>document.getElementById("input")).value;
  if (!userInput.endsWith(".zil") {
    // placeholder for future error handling
    return ;
  }
}

document.getElementById("button").addEventListener('click', () => resolve());
```

## Taking a namehash

Namehashing is an algorithm that converts a domain name in a classical format \(like example.crypto\) to a token id that Zilliqa contract can understand.

{% hint style="warning" %}
It is essential to know the difference between Zilliqa namehashing and [EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm). In ZIL we use **sha256 from SHA-2**, instead of **keccak256** which is used across the Ethereum chain
{% endhint %}

To tokenize our domain we need to split it by the "." character into separate labels, reverse the array, and reduce it to a single hash using a childhash function. Childhash function generates a hash of the current label, concatenates it with the parent, and hashes the result as a hex value. As our first parent, we take a string of **64 zeros**

```javascript

> // index.ts

Reviewer 2: @JohnnyJumper We should add imports in the index.ts

    id: "1",
    jsonrpc: "2.0",
    params
  };

  return await fetch(ZILLIQA_API, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(body),
  }).then(res => res.json());
}
```

The only parameter that is going to be changed is the params field which is nothing else but a simple array with strictly 3 arguments: **contract address**, **contract field name**, **state keys**

Let's update our resolve function and use the **fetchZilliqa** function with the following params

```typescript
// index.ts
const UD_REGISTRY_CONTRACT = "9611c53BE6d1b32058b2747bdeCECed7e1216793";

async function resolve() {
  const userInput = document.getElementById("input").value;
  if (!userInput.endsWith(".zil") {
    // placeholder for future error handeling
    return ;
  }
  const hash = namehash(userInput);
  const contractAddresses = 
    await fetchZilliqa([UD_REGISTRY_CONTRACT, "records", [hash]]);
    
  if (contractAddress.result === null) {
    // placeholder for future error handeling
    return ;
  }
  const [ownerAddress, resolverAddress] = 
    contractAddresses.result.records[hash].arguments;
  
  ...
}
```

> Calling fetchZilliqa with namehash of _**brad.zil**_ returns us the following:

Reviewer 2: How can i check this with the code outlined above?

```javascript
{
  id: "1",
  jsonrpc: "2.0",
  result: {
    records: {
        0x5fc604da00f502da70bfbc618088c0ce468ec9d18d05540935ae4118e8f50787: {
            argtypes: [],
            arguments: [
                "0x2d418942dce1afa02d0733a2000c71b371a6ac07",
                "0xdac22230adfe4601f00631eae92df6d77f054891"
            ],
            constructor: "Record"
        }
    }
  }
}
```

{% hint style="info" %}
Order is very important, as the first address in the arguments array is the owner address and the second one is a resolver contract address
{% endhint %}

{% hint style="warning" %}
Make sure to display your user an appropriate error if an owner address is not set. This means that the domain is not registered under any user and is free to be taken. 
{% endhint %}

## Fetching the domain records

As the last step we are going to make use of fetchZilliqa again, only this time we will change our params to contain the **resolver address** and for the state keys we will pass an empty array

```typescript
// index.ts resolve function
const records = await fetchZilliqa([
    resolverAddress.replace("0x", ""),
    "records",
    []
  ]);

>   console.log(records.result.records);

Reviewer 2: @JohnnyJumper How can i check the result? Should i build the project?

```

{% hint style="danger" %}

> **It is very important to remove the leading 0x from the contract address**

Reviewer 2: Please outline why it's important (Zilliqa Blockchain API requirements)

  const resolverRecord = document.createElement('span');
  resolverRecord.innerHTML = `resolverAddress: ${resolverAddress}`;

  mainContainer.appendChild(ownerRecord);
  mainContainer.appendChild(resolverRecord);

  Object.entries(records).map(([key, value]) => {
    const recordSpan = document.createElement('span');
    recordSpan.innerHTML = `${key} : ${value}`;
    mainContainer.appendChild(recordSpan);
  });
}
```

We should see something like following on successful resolution

![Screenshot](../.gitbook/assets/screen-shot-2021-01-26-at-10.33.11-pm.png)

### Error handling

Now that we have made a successful call let's deal with all possible errors that could happen during the resolution. For this purposes, we are going to create a function that will place an error in our records div

```typescript
// index.ts
function displayError(message: string, cleanDom?: boolean) {
  const mainContainer = document.getElementById('records');
  if (cleanDom) {
    cleanDOM(mainContainer);
  }
  const error = document.createElement('p');
  error.style.color = "red";
  error.innerHTML = message;
  mainContainer.appendChild(error);
  return ;
}
```

Errors that can happen are 

> * **Domain is not registered** -&gt; Happens when we couldn't get the owner address

Reviewer 2: @JohnnyJumper Please convert errors references to a small markdown table

* **Domain is not supported** -&gt; When we trying to resolve a domain that doesn't end with .zil
* **Domain is not configured** -&gt; It is possible the owner has claimed the domain but the resolver contract wasn't configured \(meaning it has the owner address and doesn't have the resolver address attached\) 
* **Record is not found** -&gt; This can be useful if you are querying domain for specific records and it is null

> Although any string can be stored as a key under the domain, Unstoppable domains [standardized some of the keys](https://docs.unstoppabledomains.com/domain-registry-essentials/records-reference) across many applications.   

Reviewer 2: @JohnnyJumper Please, follow google's cross-reference guide https://developers.google.com/style/cross-references

For the record is not found error message we are going to check if the domain has a BTC address and if not we will show the error without cleaning the entire DOM  

> 

Reviewer 2: Is that a final version of index.ts? We should outline it.

    displayError('crypto.BTC.address: Record is not found', false);
  }
}

async function resolve() {
  const userInput = (<HTMLInputElement>document.getElementById("input")).value;
  if (!userInput.endsWith(".zil")) {
    displayError('domain is not supported');
    return ;
  }

  const hash = namehash(userInput);  
  const contractAddresses = await fetchZilliqa([UD_REGISTRY_CONTRACT, "records", [hash]]);
  
  if (contractAddresses.result == null) {
    displayError('domain is not registered', true);
    return ;
  }
  const [ownerAddress, resolverAddress] = await contractAddresses.result.records[hash].arguments;

  if (resolverAddress === "0x0000000000000000000000000000000000000000") {
    displayError('domain is not configured', true);
    return ;
  }

  const records = await fetchZilliqa([
    resolverAddress.replace("0x", ""),
    "records",
    []
  ]).then(data => (data.result.records));
  
  displayResolution({resolverAddress, ownerAddress, records});
}
```

{% hint style="info" %}
If domain doesn't have a resolver address it will be set as **"0x0000000000000000000000000000000000000000"**

_You  can check it with the domain: **paulalcock.zil**_
{% endhint %}

> By this moment you can successfully resolve any .zil domain and show an appropriate error message for your users.

Reviewer 2: What to do after the code is written? How to build the project? How file tree will look after the build?

> 

Reviewer 2: Please, provide a link to the repo with the code in the and of the article.  
 
Add a link to the discord community.

* [Displaying results](resolve-zil-domain.md#displaying-result)
* [Error handling](resolve-zil-domain.md#error-handling)

## Initialize the project folder

#### Install dependencies

First, install the project's two dependencies:

1. [parcel.js](https://parceljs.org/) — A simple bundler for fast development
2. [hash.js](https://www.npmjs.com/package/hash.js/v/1.1.7) — For taking sha256 hashes of a domain name

```shell
npm install -g add parcel-bundler

mkdir unstoppable-zil-resolution
cd unstoppable-zil-resolution

npm init
npm install hash.js

touch index.ts index.html
```

Your folder structure should look like this.

```
.
├── index.html
├── index.ts
├── node_modules
├── package.json

```

#### Create `package.json`

Next, open `package.json` to configure your build scripts.

As mentioned above, this project uses `parcel.js`. One of the features of which is a built in development server. We can activate this server by calling `parcel <entry file>`

Parcel also requires us to add a `browserlist` due to its [babel technology](https://flaviocopes.com/parcel-regeneratorruntime-not-defined/).

The resulting `package.json` file should look like this.

> 

Reviewer 2: ```suggestion
#### `package.json`
```

```javascript

> // package.json

Reviewer 2: Please, move file names to headings

| **ZILLIQA_API** | an official Zilliqa endpoint which serves us an entry point to the Zilliqa blockchain |
| **UD_REGISTRY_CONTRACT_ADDRESS** | Registry address of UD without _0x_ prefix |

We'll discuss the registry contract address later in this guide.

#### Create `index.js`

```javascript
// index.js
import hash from 'hash.js';

const ZILLIQA_API = "https://api.zilliqa.com/";
const UD_REGISTRY_CONTRACT_ADDRESS = "9611c53BE6d1b32058b2747bdeCECed7e1216793";
```

Next, we need to define and attach a function `resolve` to our HTML button under `id="button"`.

We start writing this function by simply taking our input from the text field and preparing to handle an incorrect domain. We will revisit [error handling](resolve-zil-domain.md#error-handling) later in this guide.

{% hint style="warning" %}
Any domain that does not end with `.zil` is out of scope for this guide.
{% endhint }

#### Add .zil resolution

```javascript
// index.js
import hash from 'hash.js';

const ZILLIQA_API = "https://api.zilliqa.com/";
const UD_REGISTRY_CONTRACT_ADDRESS = "9611c53BE6d1b32058b2747bdeCECed7e1216793";

async function resolve() {
  const userInput = (document.getElementById("input")).value;

>   if (!userInput.endsWith(".zil") {

Reviewer: ```suggestion
  if (!userInput.endsWith(".zil")) {
```

    // placeholder for future error handling
    return;
  }
}

document.getElementById("button").addEventListener('click', () => resolve());
```

## Taking a namehash

Namehashing is an algorithm that tokenizes your domain name in a way that a Zilliqa smart contract can understand.

{% hint style="warning" %}
It is essential to know the difference between Zilliqa namehashing and [EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm). Zilliqa uses **sha256 from SHA-2**, instead of **keccak256** which is more commonly used in Ethereum.
{% endhint %}

To tokenize our domain we need to split it by the "." character into separate labels, reverse the array, and reduce it to a single hash using a childhash function. Childhash generates a hash of the current label, concatenates it with the parent, and hashes the result as a hex value. For our first parent, we'll take a string of **64 zeros**.

#### Adding a namehash

```javascript
// index.js

> function namehash(domain) {
>   const parent =
>     '0000000000000000000000000000000000000000000000000000000000000000';
>   return '0x' + [parent]
>     .concat(
>       domain
>       .split('.')
>       .reverse()
>     )
>     .reduce((parent, label) =>
>       childhash(parent, label),
>     );
> }
> 
> function childhash(parentHash, label) {
>   parentHash = parentHash.replace(/^0x/, '');
>   const labelHash = sha256(label)
>   return sha256(parentHash + labelHash, "hex");
> }
> 
> function sha256(message, inputEnc) {
>   return hash.sha256()
>     .update(message, inputEnc)
>     .digest('hex');
> }

Reviewer: ```suggestion
function namehash(name) {
    if (!name) {
        return new Uint8Array(32);
    }
    const [label, ...remainder] = name.split('.');
    const labelHash = sha256.array(label);
    const remainderHash = namehash(remainder.join('.'));
    return sha256.array(new Uint8Array([...remainderHash, ...labelHash]));
}

function arrayToHex(arr) {
    return '0x' + Array.prototype.map.call(arr, x => ('00' + x.toString(16)).slice(-2)).join('');
}
```

> function namehash(domain) {
>   const parent =
>     '0000000000000000000000000000000000000000000000000000000000000000';
>   return '0x' + [parent]
>     .concat(
>       domain
>       .split('.')
>       .reverse()
>     )
>     .reduce((parent, label) =>
>       childhash(parent, label),
>     );
> }
> 
> function childhash(parentHash, label) {
>   parentHash = parentHash.replace(/^0x/, '');
>   const labelHash = sha256(label)
>   return sha256(parentHash + labelHash, "hex");
> }
> 
> function sha256(message, inputEnc) {
>   return hash.sha256()
>     .update(message, inputEnc)
>     .digest('hex');
> }

Reviewer: ```suggestion
function namehash(name) {
    if (!name) {
        return new Uint8Array(32);
    }
    const [label, ...remainder] = name.split('.');
    const labelHash = sha256.array(label);
    const remainderHash = namehash(remainder.join('.'));
    return sha256.array(new Uint8Array([...remainderHash, ...labelHash]));
}

function arrayToHex(arr) {
    return '0x' + Array.prototype.map.call(arr, x => ('00' + x.toString(16)).slice(-2)).join('');
}
```

Reviewer: Works with `js-sha256`

Reviewer: Call it like `arrayToHex(namehash('test.zil'))`

Author: This doesn't really produce the same hash as needed.

I ran it with js-sha256 and this is the output I got
```javascript
{
oldHashFunctionWithReduce: "0x5fc604da00f502da70bfbc618088c0ce468ec9d18d05540935ae4118e8f50787",
newHashFunctionWithShaArray: "0xaf2992e5b04145f47ac8ec8b2a55c807c76929834168c3540fd1d9efedd39f1e"
}
```

I think it is because the original namehash function is going between different input encoding. At the start, it uses a normal input as a character array, and only when we hash the concatenated result we are interested in the hex input opposite to the character. The string "0xab"  produces the following hash: `"a50350de8cbfbc43ee04e13e8287390194eaf07db29ec24071f88a03ab5a8541"` while the same input as hex i.e. 0xab produces the following hash: `"0xaf2992e5b04145f47ac8ec8b2a55c807c76929834168c3540fd1d9efedd39f1e"`

Reviewer: @JohnnyJumper are you sure that you're testing the same code I've suggested? I can't reproduce the result you're getting.
Both the new and the old versions return `0x5fc604da00f502da70bfbc618088c0ce468ec9d18d05540935ae4118e8f50787` for `brad.zil` for me.
Could you push the changes that you're testing?

Author: It was a stupid mistake from my side. I used old namehash function when was testing inside the new one. didn't notice I renamed it for namehash2. My bad

#### Fetch the contract addresses

```javascript
// index.js
async function fetchZilliqa(params) {
  const body = {
    method: "GetSmartContractSubState",
    id: "1",
    jsonrpc: "2.0",
    params
  };

  return await fetch(ZILLIQA_API, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(body),
  }).then(res => res.json());
}
```

The parameters we need to send are 

* The **contract address** we want to query,
* The **contract field name** we will use string `records`,
* The **contract state keys** array of strings, in our case namehash of the domain is the only value in the array, 

Let's update our resolve function and use the **fetchZilliqa** function

#### Adding address resolution

```javascript
// index.js
async function resolve() {
  const userInput = document.getElementById("input").value;
  if (!userInput.endsWith(".zil") {
    // placeholder for future error handling
    return;
  }

  const hash = namehash(userInput);
  const registryState =
    await fetchZilliqa([UD_REGISTRY_CONTRACT_ADDRESS, "records", [hash]]);

>   if (contractAddress.result === null) {

Reviewer: `contractAddress` isn't defined in this scope

    // placeholder for future error handling
    return;
  }

>   const [ownerAddress, resolverAddress] =
>   registryState.result.records[hash].arguments; console.log({

Reviewer: ```suggestion
  const [ownerAddress, resolverAddress] = registryState.result.records[hash].arguments;
  console.log({
```
